packages/yoastseo/src/languageProcessing/researches/keywordCount.js

```diff
@@ -1,65 +1,100 @@
+		const hasAllKeywords = matchesInSentence.every( wordForms => wordForms.count > 0 );
 
 		if ( hasAllKeywords ) {
 			const counts = matchesInSentence.map( match => match.count );
+			const totalMatchCount = Math.min( ...counts );
 			const foundWords = flattenDeep( matchesInSentence.map( match => match.matches ) );
-			keywordsFound.count += Math.min( ...counts );
-			keywordsFound.matches.push( foundWords );
-			keywordsFound.sentencesWithKeywords.push( sentence );
+
+			let markings = [];
+
+			if ( matchWordCustomHelper ) {
+				// Currently, this check is only applicable for Japanese.
+				markings = markWordsInASentence( sentence.text, foundWords, matchWordCustomHelper );
+			} else {
+				markings = getMarkingsInSentence( sentence, foundWords, locale );
+			}
+
+			result.count += totalMatchCount;
+			result.markings.push( markings );
 		}
 	} );
 
-	const matches = unique( flattenDeep( keywordsFound.matches ) ).sort( ( a, b ) => b.length - a.length );
-	const keyphraseForms = flattenDeep( topicForms.keyphraseForms );
+	return result;
+}
+
+/**
+ * Calculates the keyphrase count, takes morphology into account.
+ *
+ * @param {Paper}       paper       The paper containing keyphrase and text.
+ * @param {Researcher}  researcher  The researcher.
+ *
+ * @returns {Object} An object containing an array of all the matches, markings and the keyphrase count.
+ */
+export default function keyphraseCount( paper, researcher ) {
+	const topicForms = researcher.getResearch( "morphology" );
+	topicForms.keyphraseForms = topicForms.keyphraseForms.map( word => word.map( form => normalizeSingle( form ) ) );
+
+	if ( topicForms.keyphraseForms.length === 0 ) {
+		return { count: 0, markings: [], length: 0 };
+	}
+
+	const matchWordCustomHelper = researcher.getHelper( "matchWordCustomHelper" );
+	const locale = paper.getLocale();
+	const sentences = getSentencesFromTree( paper );
+	// Exact matching is requested when the keyphrase is enclosed in double quotes.
+	const isExactMatchRequested = isDoubleQuoted( paper.getKeyword() );
+
+	/*
+	* Count the amount of keyphrase occurrences in the sentences.
+	* An occurrence is counted when all words of the keyphrase are contained within the sentence. Each sentence can contain multiple keyphrases.
+	* (e.g. "The apple potato is an apple and a potato." has two occurrences of the keyphrase "apple potato").
+	* */
+	const keyphraseFound = countKeyphraseInText( sentences, topicForms, locale, matchWordCustomHelper, isExactMatchRequested );
 
 	return {
-		count: keywordsFound.count,
-		matches: matches,
-		markings: markWordsInSentences( keyphraseForms, keywordsFound.sentencesWithKeywords, locale, matchWordCustomHelper ),
+		count: keyphraseFound.count,
+		markings: flatten( keyphraseFound.markings ),
 		length: topicForms.keyphraseForms.length,
 	};
 }
+
+/**
+ * Calculates the keyphrase count, takes morphology into account.
+ *
+ * @deprecated Since version 20.12. Use keyphraseCount instead.
```

Comment:
The version here could change depending on our plan to ship this change

Reply:
See comment above.